src/cdomains/intDomain.ml

```diff
@@ -648,12 +648,24 @@ struct
   | `Definite x -> if i = x then `Eq else `Neq
   | `Excluded (s,r) -> if S.mem i s then `Top else `Neq
 
-  let cast_to t = function
+  let top_of ik = `Excluded (S.empty (), size ik)
+  let top_if_not_in_int64 ik f x = try f x with Size.Not_in_int64 -> top_of ik
+  let cast_to ?torg ik = top_if_not_in_int64 ik @@ function
     | `Excluded (s,r) ->
-      (let r' = size t in (* target range *)
-      try `Excluded (if R.leq r r' then S.map (Integers.cast_to t) s, r' else S.empty (), r') (* TODO can we do better here? *) with
-      Size.Not_in_int64 -> top_of t)
-    | `Definite x -> (try `Definite (Integers.cast_to t x) with Size.Not_in_int64 -> top_of t)
+      let r' = size ik in
+      `Excluded (
+        if R.leq r r' then (* upcast -> no change *)
+          s, r
+        else if torg = None then (* same static type -> no overflows for r, but we need to cast s since it may be out of range after lift2_inj *)
+          let s' = S.map (Integers.cast_to ik) s in
+          s', r'
+        else (* downcast: may overflow *)
+          let s' = S.map (Integers.cast_to ik) s in
+          (* We want to filter out all i in s' where (t)x with x in r could be i. *)
+          (* Since this is hard to compute, we just keep all i in s' which overflowed, since those are safe - all i which did not overflow may now be possible due to overflow of r. *)
+          S.diff s' s, r'
```

Comment:
What if several different values in the original type (of which one was excluded) map to the same `x \in S'`? In this case we could not exclude `x`.

Are we sure this can not happen for casts in C? To me, it seems like this is bound to happen when casting something like an `unsigned long` into something much smaller such as an `unsigned char`?

I can try to come up with an example for this.

Reply:
It actually happens:
~~~c
#include "stdio.h"
int main() {
    unsigned long acc = 42;

    for(int i=0;i<10;i++) {
        unsigned char c = (char)acc;
        acc += 256;
        printf("%i\n", c);
    }
}
~~~~
Will print `42` very time.

So a case where our thing is unsound is e.g.:
~~~c
  unsigned long acc;

  if(acc != 511) {
      unsigned char c = (char)acc;
      // c → Not {255}([0,8])
  }
~~~

Reply:
What's wrong with it?

~~~c
  unsigned long ul = 511;
  char c = ul;
  unsigned char uc = ul;
  printf("%lu %d %d\n", ul, c, uc);
  // 511 -1 255
~~~

Reply:
Our analysis claims that c in the example above can not be `255`. Which is wrong.

Reply:
~~~c
  unsigned long acc = 7;

  if(top) {
    acc = 255;
  }


  if(acc != 511) {
      unsigned char c = (char)acc;
      assert(c != 255);
      int b = 42;
  }
~~~

Reply:
> Yes, but just because acc can not be 511, it does not mean that acc can not be 255.

Yes, but we're talking about `c`, not `acc`. `c` can not be 255 since that's what 511 becomes after casting.

Reply:
That's the whole point: when an excluded value overflows, it doesn't make a difference if it's excluded or definite.
It cannot be 'overwritten' by a value in the included range. The ones that don't overflow don't change, and the ones that do can't hit it since it also overflowed.
If they overflowed several times, they will also hit the same spot since `(n*t) % t` stays constant.

Reply:
Assume a concrete execution where `top` is `true`.
`acc` therefore is `255`, which is not `511`, so we go into the if branch. Casting `255` to `unsigned char` yields `255`. Therefore, `c` is `255` and our assert fails. 

The analysis says it will succeed.

Reply:
No, no better idea.

I think the only case in which we can do better than top is if we have an exclusion set that excludes all values that will map to a certain value in the smaller type. That's probably expensive to check and I also don't know when we would ever have such an exclusion set.

Reply:
I wonder if there's something one could do. Test 21/03 fails now:
~~~c
  char a;
  if(a == 127) return;
  char b = a+1; // (char)128 = -128
  assert(b != -128);
~~~